File: streamx-flink/streamx-flink-connector/streamx-flink-connector-redis/src/main/scala/com/streamxhub/streamx/flink/connector/redis/bean/RedisContainer.scala
/*
 * Copyright (c) 2019 The StreamX Project
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.streamxhub.streamx.flink.connector.redis.bean

import com.streamxhub.streamx.common.util.Logger
import org.apache.commons.pool2.impl.GenericObjectPoolConfig
import org.apache.flink.streaming.connectors.redis.common.config.{FlinkJedisConfigBase, FlinkJedisPoolConfig, FlinkJedisSentinelConfig}
import org.apache.flink.streaming.connectors.redis.common.container.{RedisContainer => BahirRedisContainer}
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommand._
import redis.clients.jedis.{Jedis, JedisPool, JedisSentinelPool}

class RedisContainer(container: BahirRedisContainer) {

  def open(): Unit = {
    container.open()
  }

  lazy val jedis: Jedis = {
    val method = container.getClass.getDeclaredMethod("getInstance")
    method.setAccessible(true)
    method.invoke(container).asInstanceOf[Jedis]
  }

  def invoke[T](mapper: RedisMapper[T], input: T, transaction: Option[redis.clients.jedis.Transaction]): Unit = {
    val key = mapper.getKeyFromData(input)
    val value = mapper.getValueFromData(input)
    mapper.getCommandDescription.getCommand match {
      case  RPUSH => transaction match {
        case Some(t) => t.rpush(key, value)
        case _ => this.container.rpush(key, value)
      }
      case LPUSH => transaction match {
        case Some(t) => t.lpush(key, value)
        case _ => this.container.lpush(key, value)
      }
      case SADD => transaction match {
        case Some(t) => t.sadd(key, value)
        case _ => this.container.sadd(key, value)
      }
      case SET => transaction match {
        case Some(t) => t.set(key, value)
        case _ => this.container.set(key, value)
      }
      case PFADD => transaction match {
        case Some(t) => t.pfadd(key, value)
        case _ => this.container.pfadd(key, value)
      }
      case PUBLISH => transaction match {
        case Some(t) => t.publish(key, value)
        case _ => this.container.publish(key, value)
      }
      case ZADD => transaction match {
        case Some(t) => t.zadd(mapper.getCommandDescription.getAdditionalKey, value.toDouble, key)
        case _ => this.container.zadd(mapper.getCommandDescription.getAdditionalKey, value, key)
      }
      case ZREM => transaction match {
        case Some(t) => t.zrem(mapper.getCommandDescription.getAdditionalKey, key)
        case _ => this.container.zrem(mapper.getCommandDescription.getAdditionalKey, key)
      }
      case HSET => transaction match {
        case Some(t) => t.hset(mapper.getCommandDescription.getAdditionalKey, key, value)
        case _ => this.container.hset(mapper.getCommandDescription.getAdditionalKey, key, value)
      }
      case other => throw new IllegalArgumentException("[StreamX] RedisSink:Cannot process such data type: " + other)
    }
  }

  def expire(key: String, ttl: Int): Unit = {
    ttl match {
      case Int.MaxValue =>
      case _ => jedis.expire(key, ttl)
    }
  }

  def close(): Unit = {
    container.close()
  }

}

object RedisContainer extends Logger {

  def getContainer(jedisConfig: FlinkJedisConfigBase): RedisContainer = {
    val genericObjectPoolConfig = new GenericObjectPoolConfig
    genericObjectPoolConfig.setMaxIdle(jedisConfig.getMaxIdle)
    genericObjectPoolConfig.setMaxTotal(jedisConfig.getMaxTotal)
    genericObjectPoolConfig.setMinIdle(jedisConfig.getMinIdle)
    try {
      val bahirRedisContainer = jedisConfig match {
        case jedisPoolConfig: FlinkJedisPoolConfig =>
          val jedisPool = new JedisPool(
            genericObjectPoolConfig,
            jedisPoolConfig.getHost,
            jedisPoolConfig.getPort,
            jedisPoolConfig.getConnectionTimeout,
            jedisPoolConfig.getPassword,
            jedisPoolConfig.getDatabase
          )
          new BahirRedisContainer(jedisPool)
        case _ =>
          val jedisSentinelConfig = jedisConfig.asInstanceOf[FlinkJedisSentinelConfig]
          val jedisSentinelPool = new JedisSentinelPool(jedisSentinelConfig.getMasterName,
            jedisSentinelConfig.getSentinels,
            genericObjectPoolConfig,
            jedisSentinelConfig.getSoTimeout,
            jedisSentinelConfig.getPassword,
            jedisSentinelConfig.getDatabase
          )
          new BahirRedisContainer(jedisSentinelPool)
      }
      val redisContainer = new RedisContainer(bahirRedisContainer)
      redisContainer.open()
      redisContainer
    } catch {
      case e: Exception =>
        logError("RedisSink:Redis has not been properly initialized: ", e)
        throw e
    }
  }

}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-redis/src/main/java/org/apache/streampark/flink/connector/redis/bean/RedisContainer.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.redis.bean;

import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisConfigBase;
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisPoolConfig;
import org.apache.flink.streaming.connectors.redis.common.config.FlinkJedisSentinelConfig;
import org.apache.flink.streaming.connectors.redis.common.mapper.RedisCommand;

import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisSentinelPool;
import redis.clients.jedis.Transaction;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.Optional;

/** Redis container wrapper with command dispatch. */
public class RedisContainer {

    private static final Logger LOG = LoggerFactory.getLogger(RedisContainer.class);

    private final org.apache.flink.streaming.connectors.redis.common.container.RedisContainer container;

    public RedisContainer(org.apache.flink.streaming.connectors.redis.common.container.RedisContainer container) {
        this.container = container;
    }

    public void open() {
        try {
            container.open();
        } catch (Exception e) {
            throw new RuntimeException("Failed to open Redis container", e);
        }
    }

    public Jedis getJedis() {
        try {
            Method method = container.getClass().getDeclaredMethod("getInstance");
            method.setAccessible(true);
            return (Jedis) method.invoke(container);
        } catch (ReflectiveOperationException e) {
            throw new IllegalStateException("Failed to obtain Jedis instance from RedisContainer", e);
        }
    }

    public <T> void invoke(RedisMapper<T> mapper, T input, Optional<Transaction> transaction) {
        String key = mapper.getKeyFromData(input);
        String value = mapper.getValueFromData(input);
        RedisCommand command = mapper.getCommandDescription().getCommand();
        switch (command) {
            case RPUSH:
                if (transaction.isPresent()) {
                    transaction.get().rpush(key, value);
                } else {
                    container.rpush(key, value);
                }
                break;
            case LPUSH:
                if (transaction.isPresent()) {
                    transaction.get().lpush(key, value);
                } else {
                    container.lpush(key, value);
                }
                break;
            case SADD:
                if (transaction.isPresent()) {
                    transaction.get().sadd(key, value);
                } else {
                    container.sadd(key, value);
                }
                break;
            case SET:
                if (transaction.isPresent()) {
                    transaction.get().set(key, value);
                } else {
                    container.set(key, value);
                }
                break;
            case PFADD:
                if (transaction.isPresent()) {
                    transaction.get().pfadd(key, value);
                } else {
                    container.pfadd(key, value);
                }
                break;
            case PUBLISH:
                if (transaction.isPresent()) {
                    transaction.get().publish(key, value);
                } else {
                    container.publish(key, value);
                }
                break;
            case ZADD:
                if (transaction.isPresent()) {
                    transaction
                            .get()
                            .zadd(mapper.getCommandDescription().getAdditionalKey(), Double.parseDouble(value), key);
                } else {
                    container.zadd(mapper.getCommandDescription().getAdditionalKey(), value, key);
                }
                break;
            case ZREM:
                if (transaction.isPresent()) {
                    transaction.get().zrem(mapper.getCommandDescription().getAdditionalKey(), key);
                } else {
                    container.zrem(mapper.getCommandDescription().getAdditionalKey(), key);
                }
                break;
            case HSET:
                if (transaction.isPresent()) {
                    transaction.get().hset(mapper.getCommandDescription().getAdditionalKey(), key, value);
                } else {
                    container.hset(
                            mapper.getCommandDescription().getAdditionalKey(), key, value, Integer.MAX_VALUE);
                }
                break;
            default:
                throw new IllegalArgumentException(
                        "[StreamPark] RedisSink:Cannot process such data type: " + command);
        }
    }

    public void expire(String key, int ttl) {
        if (ttl != Integer.MAX_VALUE) {
            getJedis().expire(key, ttl);
        }
    }

    public void close() {
        try {
            container.close();
        } catch (IOException e) {
            throw new RuntimeException("Failed to close Redis container", e);
        }
    }

    public static RedisContainer getContainer(FlinkJedisConfigBase jedisConfig) {
        GenericObjectPoolConfig<?> genericObjectPoolConfig = new GenericObjectPoolConfig<>();
        genericObjectPoolConfig.setMaxIdle(jedisConfig.getMaxIdle());
        genericObjectPoolConfig.setMaxTotal(jedisConfig.getMaxTotal());
        genericObjectPoolConfig.setMinIdle(jedisConfig.getMinIdle());
        try {
            org.apache.flink.streaming.connectors.redis.common.container.RedisContainer bahirRedisContainer;
            if (jedisConfig instanceof FlinkJedisPoolConfig) {
                FlinkJedisPoolConfig jedisPoolConfig = (FlinkJedisPoolConfig) jedisConfig;
                JedisPool jedisPool =
                        new JedisPool(
                                genericObjectPoolConfig,
                                jedisPoolConfig.getHost(),
                                jedisPoolConfig.getPort(),
                                jedisPoolConfig.getConnectionTimeout(),
                                jedisPoolConfig.getPassword(),
                                jedisPoolConfig.getDatabase());
                bahirRedisContainer =
                        new org.apache.flink.streaming.connectors.redis.common.container.RedisContainer(jedisPool);
            } else {
                FlinkJedisSentinelConfig jedisSentinelConfig = (FlinkJedisSentinelConfig) jedisConfig;
                JedisSentinelPool jedisSentinelPool =
                        new JedisSentinelPool(
                                jedisSentinelConfig.getMasterName(),
                                jedisSentinelConfig.getSentinels(),
                                genericObjectPoolConfig,
                                jedisSentinelConfig.getSoTimeout(),
                                jedisSentinelConfig.getPassword(),
                                jedisSentinelConfig.getDatabase());
                bahirRedisContainer =
                        new org.apache.flink.streaming.connectors.redis.common.container.RedisContainer(
                                jedisSentinelPool);
            }
            RedisContainer redisContainer = new RedisContainer(bahirRedisContainer);
            redisContainer.open();
            return redisContainer;
        } catch (Exception e) {
            LOG.error("RedisSink:Redis has not been properly initialized: ", e);
            throw e;
        }
    }
}
